ocaml/tests/test_vm_helpers.ml

```diff
@@ -573,6 +609,244 @@ let test_get_group_key_vgpu_and_netsriov () =
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"; "h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(1) h2(1) h3(1))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h1"); ("vm2", "h2"); ("vm3", "h3")]
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"; "h2"; "h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(0) h2(1) h3(2))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h2"); ("vm2", "h3"); ("vm3", "h3")]
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"]; ["h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 starting VMs (h1(0) h2(0) h3(3))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h3"); ("vm2", "h3"); ("vm3", "h3")]
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"; "h2"]; ["h3"]]
+    }
+  ; {
+      description= "3 running VMs with home server"
+    ; state= "running"
+    ; vms= ["vm1"; "vm2"; "vm3"]
+    ; resident_on= [("vm1", "h1"); ("vm2", "h2"); ("vm3", "h3")]
+    ; group= "anti_affinity"
+    ; home_server= Some "h3"
+    ; expected= [["h3"]; ["h1"; "h2"]]
+    }
+  ; {
+      description= "6 running VMs (h1(2) h2(2) h3(2))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"; "vm4"; "vm5"; "vm6"]
+    ; resident_on=
+        [
+          ("vm1", "h1")
+        ; ("vm2", "h1")
+        ; ("vm3", "h2")
+        ; ("vm4", "h2")
+        ; ("vm5", "h3")
+        ; ("vm6", "h3")
+        ]
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"; "h2"; "h3"]]
+    }
+  ; {
+      description= "6 running VMs (h1(1) h2(2) h3(3))"
+    ; state= "starting"
+    ; vms= ["vm1"; "vm2"; "vm3"; "vm4"; "vm5"; "vm6"]
+    ; resident_on=
+        [
+          ("vm1", "h1")
+        ; ("vm2", "h2")
+        ; ("vm3", "h2")
+        ; ("vm4", "h3")
+        ; ("vm5", "h3")
+        ; ("vm6", "h3")
+        ]
+    ; group= "anti_affinity"
+    ; home_server= None
+    ; expected= [["h1"]; ["h2"]; ["h3"]]
+    }
+  ]
+
+let test_anti_affinity
+    {state; vms; resident_on; group; home_server; expected; _} () =
+  let __context = T.make_test_database () in
+  let host1 = List.hd (Db.Host.get_all ~__context) in
+  let host2 = T.make_host ~__context () in
+  let host3 = T.make_host ~__context () in
+  let host_map = [("h1", host1); ("h2", host2); ("h3", host3)] in
+  let vm = T.make_vm ~__context () in
+  let placement =
+    match group with "anti_affinity" -> `anti_affinity | _ -> `normal
+  in
+  let group = T.make_vm_group ~__context ~placement () in
+  Db.VM.set_groups ~__context ~self:vm ~value:[group] ;
+  vms
+  |> List.map (fun key ->
+         let vm = T.make_vm ~__context () in
+         (key, vm)
+     )
+  |> List.iter (fun (key, vm) ->
+         Db.VM.set_groups ~__context ~self:vm ~value:[group] ;
```

Comment:
Seems all VMs are in the same anti-affinity group?
Then we will need more tests for:
1. with VMs not in anti-affinity group
2. with VMs in other anti-affinity group